File: language/src/main/java/com/oracle/truffle/lama/LamaLanguage.java
package com.oracle.truffle.lama;

import com.oracle.truffle.api.*;
import com.oracle.truffle.api.TruffleLanguage.ContextPolicy;
import com.oracle.truffle.api.debug.DebuggerTags;
import com.oracle.truffle.api.dsl.NodeFactory;
import com.oracle.truffle.api.instrumentation.ProvidedTags;
import com.oracle.truffle.api.instrumentation.StandardTags;
import com.oracle.truffle.api.interop.InteropLibrary;
import com.oracle.truffle.api.nodes.Node;
import com.oracle.truffle.api.nodes.NodeInfo;
import com.oracle.truffle.api.nodes.RootNode;
import com.oracle.truffle.api.source.Source;
import com.oracle.truffle.lama.builtins.LamaBuiltinNode;
import com.oracle.truffle.lama.parser.LamaLanguageParser;
import com.oracle.truffle.lama.runtime.LamaContext;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

@TruffleLanguage.Registration(id = LamaLanguage.ID, name = "Lama", defaultMimeType = LamaLanguage.MIME_TYPE, characterMimeTypes = LamaLanguage.MIME_TYPE, contextPolicy = ContextPolicy.SHARED, fileTypeDetectors = LamaFileDetector.class)
@ProvidedTags({StandardTags.CallTag.class, StandardTags.StatementTag.class, StandardTags.RootTag.class, StandardTags.RootBodyTag.class, StandardTags.ExpressionTag.class, DebuggerTags.AlwaysHalt.class, StandardTags.ReadVariableTag.class, StandardTags.WriteVariableTag.class})
public final class LamaLanguage extends TruffleLanguage<LamaContext> {
    public static volatile int counter;

    public static final String ID = "lama";
    public static final String MIME_TYPE = "application/x-lama";
    private static final Source BUILTIN_SOURCE = Source.newBuilder(LamaLanguage.ID, "", "Lama builtin").build();

    private final Assumption singleContext = Truffle.getRuntime().createAssumption("Single Lama context.");

    private final Map<NodeFactory<? extends LamaBuiltinNode>, RootCallTarget> builtinTargets = new ConcurrentHashMap<>();
    private final Map<String, RootCallTarget> undefinedFunctions = new ConcurrentHashMap<>();

    public LamaLanguage() {
        counter++;
    }

    @Override
    protected LamaContext createContext(Env env) {
        return new LamaContext(this, env);
    }

    @Override
    protected boolean patchContext(LamaContext context, Env newEnv) {
        context.patchContext(newEnv);
        return true;
    }


    public static NodeInfo lookupNodeInfo(Class<?> clazz) {
        if (clazz == null) {
            return null;
        }
        NodeInfo info = clazz.getAnnotation(NodeInfo.class);
        if (info != null) {
            return info;
        } else {
            return lookupNodeInfo(clazz.getSuperclass());
        }
    }

    @Override
    protected CallTarget parse(ParsingRequest request) throws Exception {
        Source source = request.getSource();
        RootNode evalMain = LamaLanguageParser.parseLama(this, source);
        return Truffle.getRuntime().createCallTarget(evalMain);
    }

    @Override
    protected void initializeMultipleContexts() {
        singleContext.invalidate();
    }


    @Override
    protected boolean isVisible(LamaContext context, Object value) {
        return !InteropLibrary.getFactory().getUncached(value).isNull(value);
    }


    private static final LanguageReference<LamaLanguage> REFERENCE = LanguageReference.create(LamaLanguage.class);

    public static LamaLanguage get(Node node) {
        return REFERENCE.get(node);
    }

    private static final List<NodeFactory<? extends LamaBuiltinNode>> EXTERNAL_BUILTINS = Collections.synchronizedList(new ArrayList<>());

    public static void installBuiltin(NodeFactory<? extends LamaBuiltinNode> builtin) {
        EXTERNAL_BUILTINS.add(builtin);
    }

}

File: language/src/main/java/com/oracle/truffle/lama/LamaLanguage.kt
package com.oracle.truffle.lama

import com.oracle.truffle.api.CallTarget
import com.oracle.truffle.api.Truffle
import com.oracle.truffle.api.TruffleLanguage
import com.oracle.truffle.api.TruffleLanguage.ContextPolicy
import com.oracle.truffle.api.TruffleLanguage.Registration
import com.oracle.truffle.api.debug.DebuggerTags
import com.oracle.truffle.api.instrumentation.ProvidedTags
import com.oracle.truffle.api.instrumentation.StandardTags.*
import com.oracle.truffle.api.interop.InteropLibrary
import com.oracle.truffle.api.nodes.Node
import com.oracle.truffle.api.nodes.NodeInfo
import com.oracle.truffle.api.nodes.RootNode
import com.oracle.truffle.lama.runtime.LamaContext

@Registration(
    id = LamaLanguage.ID,
    name = "Lama",
    defaultMimeType = LamaLanguage.MIME_TYPE,
    characterMimeTypes = [LamaLanguage.MIME_TYPE],
    contextPolicy = ContextPolicy.SHARED,
    fileTypeDetectors = [SLFileDetector::class]
)
@ProvidedTags(
    CallTag::class,
    StatementTag::class,
    RootTag::class,
    RootBodyTag::class,
    ExpressionTag::class,
    DebuggerTags.AlwaysHalt::class,
    ReadVariableTag::class,
    WriteVariableTag::class
)
class LamaLanguage : TruffleLanguage<LamaContext>() {
    private val singleContext = Truffle.getRuntime().createAssumption("Single lama context.")

    override fun createContext(env: Env): LamaContext {
        return LamaContext(this, env)
    }

    override fun patchContext(context: LamaContext, newEnv: Env): Boolean {
        context.patchContext(newEnv)
        return true
    }

    override fun parse(request: ParsingRequest): CallTarget {
//        val evalMain: RootNode = SimpleLanguageParser.parseSL(this, request.source) as RootNode
        val evalMain: RootNode? = null
        return Truffle.getRuntime().createCallTarget(evalMain)
    }

    override fun initializeMultipleContexts() {
        singleContext.invalidate()
    }

    fun isSingleContext(): Boolean {
        return singleContext.isValid
    }

    override fun isVisible(context: LamaContext, value: Any): Boolean {
        return !InteropLibrary.getFactory().getUncached(value).isNull(value)
    }

    companion object {
        const val ID = "lama"
        const val MIME_TYPE = "application/x-lama"

        @Volatile
        var counter = 0
        fun lookupNodeInfo(clazz: Class<*>?): NodeInfo? {
            if (clazz == null) {
                return null
            }
            val info = clazz.getAnnotation(NodeInfo::class.java)
            return info ?: lookupNodeInfo(clazz.superclass)
        }

        private val REFERENCE: LanguageReference<LamaLanguage> = LanguageReference.create(LamaLanguage::class.java)
        operator fun get(node: Node?): LamaLanguage {
            return REFERENCE[node]
        }
    }

    init {
        counter++
    }
}